R/SimulateSOI.R

```diff
@@ -0,0 +1,246 @@
+                empty_count <- I - sum(com)
+                counts <- c(com, empty_count)
+            } else {
+                stop("com vector argument can only be either of length N, or N+1
+                    in case the number of empty sites are included")
+            }
+
+            # making sure counts vector adds up to I with the empty sites
+            # (if not, the difference is added or subtracted from the empty
+            # sites to get a total of I)
+            if(sum(counts)!=I){
+                counts[N+1] <- counts[N+1] + (I-sum(counts))
+            }
+
+            # TRANSITION MATRIX & INTERACTION RATES
+            # Initialise trans_mat transition matrix reprenting the actual
+            # reactions
+            # First N columns = migration: species_i counts go up with 1,
+            # empty sites go down with -1
+            # Next N columns = death: species_i counts go down with -1,
+            # empty sites go up with +1
+            # Last columns: the competition / negative interaction jumps:
+            trans_mat <- cbind(
+                rbind(diag(1, ncol = N, nrow = N), rep(-1, times = N)),
+                rbind(diag(-1, ncol = N, nrow = N), rep(1, times = N))
+            )
+
+            # Interaction:
+            # happens when happens when sp_stronger interacts more strongle
+            # with sp_weaker, than sp_weaker with sp_stronger
+
+            pos_inter_rates <- matrix(0, nrow = N, ncol = N)
+            neg_inter_rates <- pos_inter_rates # copy to initialise
+
+            # pos interaction and immigration
+            # AND sp_stronger interacts positively with sp_weaker
+            for(stronger in 1:N){
+                weaker <- which(A[stronger,] > A[,stronger] & A[,stronger] >= 0)
+            if(length(weaker) > 0){
+                rate <- A[stronger, weaker] + A[weaker, stronger]
+                pos_inter_rates[stronger,weaker] <- rate
+                }
+            }
+
+            # neg interaction: competition
+            # AND sp_weaker interacts negatively with sp_stronger
+            for(stronger in 1:N){
+            weaker_vec <- which(A[stronger,] > A[,stronger] & A[,stronger] < 0)
+                if(length(weaker_vec) > 0){
+                rate <- A[stronger, weaker_vec] - A[weaker_vec, stronger]
+                neg_inter_rates[stronger, weaker_vec] <- rate
+                for(weaker in weaker_vec){
+                    jump <- rep(0, times = N+1)
+                    jump[stronger] <- 1
+                    jump[weaker] <- -1
+                    trans_mat <- cbind(trans_mat, jump)
+                    }
+                }
+            }
+
+            # INITIAL PROPENSITIES
+        propensities <- update_propensities(I, counts, death_rates, migr_rates,
+                                            pos_inter_rates, neg_inter_rates)
+
+            # K-LEAPS SIMULATION
+            sys_t <- seq(from = 0, to = tend, length.out = tend)
+
```

Comment:
Why we have indentation here suddenly changing? not good for readability of the code

Reply:
It was because I avoid exceeding character length limit in a line but now I fixed it